phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java

```
@@ -931,220 +951,298 @@ private void send(Iterator<TableRef> tableRefIterator) throws SQLException {
-                                phoenixTransactionContext.markDMLFence(table);
-                            }
-                            // Only pass true for last argument if the index is being written to on it's own (i.e. initial
-                            // index population), not if it's being written to for normal maintenance due to writes to
-                            // the data table. This case is different because the initial index population does not need
-                            // to be done transactionally since the index is only made active after all writes have
-                            // occurred successfully.
-                            hTable = phoenixTransactionContext.getTransactionalTableWriter(connection, table, hTable, tableInfo.isDataTable() && table.getType() == PTableType.INDEX);
+
+            Map<TableInfo, List<Mutation>> unverifiedIndexMutations = new LinkedHashMap<>();
+            Map<TableInfo, List<Mutation>> verifiedOrDeletedIndexMutations = new LinkedHashMap<>();
+            filterIndexCheckerMutations(physicalTableMutationMap, unverifiedIndexMutations,
+                    verifiedOrDeletedIndexMutations);
+
+            // Phase 1: Send verified indexes with VERIFIED=false
+            // addRowMutations generates the mutations with VERIFIED=false
+            sendMutations(unverifiedIndexMutations.entrySet().iterator(), span, indexMetaDataPtr);
+
+            // Phase 2: Send data table and other indexes
+            Iterator<Entry<TableInfo, List<Mutation>>>
+                    mutationsIterator =
+                    physicalTableMutationMap.entrySet().iterator();
+            sendMutations(mutationsIterator, span, indexMetaDataPtr);
+
+            // Phase 3: Send verfied indexes with VERIFIED = true
+            try {
+                sendMutations(verifiedOrDeletedIndexMutations.entrySet().iterator(), span, indexMetaDataPtr);
+            } catch (SQLException ex) {
+                LOGGER.warn(
```

Please add a TODO to emit metric here. Ignoring the exception may be fine, but it would be good to track how often this happens since it has implications on read-repairs.

---

ok